pandablocks/asyncio.py

```diff
@@ -142,35 +141,41 @@ async def data(
                 `asyncio.TimeoutError`
         """
 
-        data_stream = _StreamHelper()
-        await data_stream.connect(self._host, 8889),
-
+        stream = _StreamHelper()
         connection = DataConnection()
-        data: Deque[Data] = deque()
-        reader = data_stream.reader
-        # Should we flush every FrameData?
-        flush_every_frame = flush_period is None
+        queue: asyncio.Queue[Iterable[Data]] = asyncio.Queue()
 
-        async def queue_flushed_data():
-            data.extend(connection.flush())
+        def raise_timeouterror():
+            raise asyncio.TimeoutError(f"No data received for {frame_timeout}s")
+            yield
 
         async def periodic_flush():
-            if not flush_every_frame:
+            if flush_period is not None:
                 while True:
                     # Every flush_period seconds flush and queue data
-                    await asyncio.gather(
-                        asyncio.sleep(flush_period), queue_flushed_data()
-                    )
+                    await asyncio.sleep(flush_period)
+                    queue.put_nowait(connection.flush())
 
-        flush_task = asyncio.create_task(periodic_flush())
+        async def read_from_stream():
+            reader = stream.reader
+            # Should we flush every FrameData?
+            flush_every_frame = flush_period is None
+            while True:
+                try:
+                    recv = await asyncio.wait_for(reader.read(4096), frame_timeout)
+                except asyncio.TimeoutError:
+                    queue.put_nowait(raise_timeouterror())
+                    break
+                else:
+                    queue.put_nowait(connection.receive_bytes(recv, flush_every_frame))
+
+        await stream.connect(self._host, 8889)
+        await stream.write_and_drain(connection.connect(scaled))
+        fut = asyncio.gather(periodic_flush(), read_from_stream())
         try:
-            await data_stream.write_and_drain(connection.connect(scaled))
             while True:
-                received = await asyncio.wait_for(reader.read(4096), frame_timeout)
-                for d in connection.receive_bytes(received, flush_every_frame):
-                    data.append(d)
-                while data:
-                    yield data.popleft()
+                for data in await queue.get():
+                    yield data
         finally:
-            flush_task.cancel()
-            await data_stream.close()
+            fut.cancel()
```

Comment:
Adding the following code after this line seems to catch the exception and prevent traceback from being printed
```suggestion
            fut.cancel()
            try:                                                                                                                                                                
                await fut
            except asyncio.CancelledError:
                pass
```

Reply:
May be there is a better way to capture the exception

Reply:
Thanks, I've applied your idea (but using `contextlib.suppress`)